### PR TITLE
Add pre- and postprocessing, merge configuration of data and benchmarks

### PR DESCRIPTION
**Pull request by dlunde**
This proposal (draft) PR adds a set of additional features to the framework. In particular:
- Makes it possible to add pre- and postprocessing steps to benchmarks.
- Merges configuration files for data and benchmarks, making specifying data a bit more flexible (I think). Example of mixed inputs (data), pre, and post steps:
```
timing = "complete"

[pre]
runtime = "OCaml"
argument = "my-pre-step"

[[post]]
runtime = "MExpr"
argument = "my-post-step"

[[post]]
runtime = "Python"
argument = "my-other-post-step"

[[input]]
tags = ["foo", "number"]
data = "1"

[[input]]
tags = ["bar", "number"]
file = "datasets/2.txt"
```
- Makes it possible to define multiple experiments within the same directory. This is useful if, e.g., you want to vary the experiment in some compile/run option rather than through different inputs. Example:
```
[[app]]
runtime = "WebPPL"
argument = "crbd.wppl"
options = "--particles 10"

[[app]]
runtime = "WebPPL"
argument = "crbd.wppl"
options = "--particles 100"

[[app]]
runtime = "WebPPL"
argument = "crbd.wppl"
options = "--particles 1000"

[[app]]
runtime = "WebPPL"
argument = "crbd.wppl"
options = "--particles 10000"

```

@lingmar  I think you should have a look at this PR before we merge it, since I have changed many parts of the original implementation. Also, I have only tested running the experiments under `ppl` and `test`, because I do not believe I have all prerequisites installed for `tsp` and `mcore-ocaml`. Could you check that this still works as expected?
Also, I have not made any changes to `plot.mc` and `post-process.mc`. I guessed you might want to refactor this into a postprocessing step after this PR.

I have not yet updated the README, but will do so if we decide to merge this.

A general comment for the future: Working with TOML through Python in MCore is quite horrible (or, rather, working with MCore records as dictionaries/maps is quite horrible). If we want to use this framework actively for our future papers, and continue development on it, we really need a proper tool for handling TOML parsing and writing.

### CODE REVIEW

File: tool/tool/config-scanner.mc [hunk 1]
```diff
@@ -48,179 +67,188 @@ let dirRuntime = lam path.
...
-      getTiming raw
-    , runtime =
-      let r = optionGetOrElse (lam. error "expected runtime") b.runtime in
-      mapLookupOrElse (lam. error (concat "Undefined runtime: " r)) r runtimes;
-      r
-    , argument = optionGetOrElse (lam. error "expected argument") b.argument
-    , cwd = cwd
-    , data = b.data
-    }
+  lam pb : PartialBench.
```
Comment (lingmar) [inline tool/tool/config-scanner.mc:123]:
```suggestion
  lam pb : PartialBenchmark.
```

File: tool/tool/runner.mc [hunk 2]
```diff
@@ -4,24 +4,35 @@ include "config-scanner.mc"
...
                , warmups : Int
                }
 
 -- Used as dummy data object for benchmarks without data. We could use option
 -- type in result type, but this simplifies writing to toml.
-let dataEmpty : Data = {runtime = "", argument = "", cwd = "", tags = []}
-
-let insertArg = lam cmd. lam arg.
-  strReplace cmd "{argument}" arg
+let inputEmpty : Data = {runtime = "", argument = "", cwd = "", tags = []}
```
Comment (lingmar) [inline tool/tool/runner.mc:24]:
```suggestion
let inputEmpty : Input = {file = None (), data = Some "", tags = [], cwd = ""}
```

File: benchmark-suite/benchmarks/tsp/mcore/simulated-annealing/config.toml [hunk 3]
```diff
@@ -1 +1,3 @@
+[app]
```
Comment (lingmar) [inline benchmark-suite/benchmarks/tsp/mcore/simulated-annealing/config.toml:1]:
```suggestion
[[app]]
```

**Comment by dlunde** [comment]
Right, `App` stands for Application. It is used for the common structure (runtime + arguments + options + ...) of `app`, `pre`, and `post`. Suggestions for a better name are welcome (I just chose the first thing that came to mind).

**Pull request merged**